### PR TITLE
8241423: NUMA APIs fail to work in dockers due to dependent syscalls are disabled by default

### PR DESCRIPTION
**Pull request by DamonFool**
Hi all,

NUMA APIs fail to work in dockers due to dependent syscalls are disabled by default.

NUMA APIs depend on several syscalls.
E.g., `get_mempolicy` is required for `numa_get_membind` and `numa_get_interleave_mask`.
But these dependent syscalls can be unsupported for various reasons.
Especially in dockers, `get_mempolicy` is not allowed with the default configuration [1].
So it's necessary to check whether the syscalls are available.

In theory, all NUMA-related syscalls should be checked.
But it seems hard to do so because some of them like `mbind` would cause unexpected side effects.
So only `get_mempolicy` is checked in the fix, which is already enough for all the default dockers.
And this can be refined in the future if it turns out to be a problem.

Thanks.
Best regards,
Jie

[1] https://docs.docker.com/engine/security/seccomp/

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8241423](https://bugs.openjdk.java.net/browse/JDK-8241423): NUMA APIs fail to work in dockers due to dependent syscalls are disabled by default

### Reviewers
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Per Liden](https://openjdk.java.net/census#pliden) (@pliden - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4205/head:pull/4205` \
`$ git checkout pull/4205`

Update a local copy of the PR: \
`$ git checkout pull/4205` \
`$ git pull https://git.openjdk.java.net/jdk pull/4205/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4205`

View PR using the GUI difftool: \
`$ git pr show -t 4205`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4205.diff">https://git.openjdk.java.net/jdk/pull/4205.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/os/linux/os_linux.cpp
Comment (dholmes-ora) [inline src/hotspot/os/linux/os_linux.cpp:4482]: This can just be a static function in the file rather then being in the os::Linux "namespace"

File: src/hotspot/os/linux/os_linux.cpp
Comment (dholmes-ora) [inline src/hotspot/os/linux/os_linux.cpp:4489]: Is this SYS_get_mempolicy symbol guaranteed to be available on our supported Linux versions? See earlier in the file for how we handle SYS_gettid and SYS_getcpu.

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
Hi Jie,

This wasn't as bad as I thought it might be. :) But I have one suggested change and a query.

Thanks,
David

File: src/hotspot/os/linux/os_linux.cpp
Comment (DamonFool) [reply src/hotspot/os/linux/os_linux.cpp]: > This can just be a static function in the file rather then being in the os::Linux "namespace"

Updated.
Thanks.

File: src/hotspot/os/linux/os_linux.cpp
Comment (DamonFool) [reply src/hotspot/os/linux/os_linux.cpp]: > Is this SYS_get_mempolicy symbol guaranteed to be available on our supported Linux versions? See earlier in the file for how we handle SYS_gettid and SYS_getcpu.

I think it's OK to use the symbol directly since we have used it in ZGC [1].
Thanks.

[1] https://github.com/openjdk/jdk/blob/master/src/hotspot/os/linux/gc/z/zSyscall_linux.cpp#L39

**Comment by pliden** [comment]
Wasn't this fixed in https://github.com/openjdk/jdk/pull/3704? It seems we now have a similar check in two different places?

**Comment by DamonFool** [comment]
> Wasn't this fixed in #3704? 

No.
#3704 just fix the zgc crash.
Other gcs still suffer from it.

>It seems we now have a similar check in two different places?

To fix the zgc crash, we only need to check `get_mempolicy`.
But `numa_syscall_check` may be extended to other syscalls in the future.

**Comment by pliden** [comment]
But then we should remove the check in ZGC, since it serves no purpose anymore when the same check is done in `libnuma_init`.

**Comment by DamonFool** [comment]
> But then we should remove the check in ZGC, since it serves no purpose anymore when the same check is done in `libnuma_init`.

Yes, this is a more general solution.

I used to send the RFR in 2020 [1].
But the community didn't pay much attention to it at that time.
So we have to get #3704 fixed because ZGC always crashed in docker with numa, while others wouldn't.

This week, @tschatzl encouraged me to try JDK-8241423 again.
So I sent this PR yesterday hoping that it can be fixed for other gcs too.

Changes in ZGC (JDK-8241354 and JDK-8266217) would become unnecessary if this PR is accepted.
And to make it to be more friendly to get reviewed, it would be better to file another RFE to remove them in ZGC.
Thanks.

[1] https://mail.openjdk.java.net/pipermail/hotspot-dev/2020-March/041149.html

**Review by dholmes-ora — APPROVED** [review]
LGTM.

Thanks,
David

**Comment by DamonFool** [comment]
> LGTM.
> 
> Thanks,
> David

Thanks @dholmes-ora .

**Comment by pliden** [comment]
> Changes in ZGC (JDK-8241354 and JDK-8266217) would become unnecessary if this PR is accepted.
> And to make it to be more friendly to get reviewed, it would be better to file another RFE to remove them in ZGC.

I just think the backout of JDK-8241354 and JDK-8266217 should have have been part of this PR. Otherwise this gets harder to review, since the question why we do redundant checks now pops up. But sure, if you prefer backing out JDK-8241354 and JDK-8266217 as a separate change, I'm fine with that too.

**Comment by DamonFool** [comment]
> But sure, if you prefer backing out JDK-8241354 and JDK-8266217 as a separate change, I'm fine with that too.

Thanks @pliden .
I'll file a RFE after this pr.

**Comment by DamonFool** [comment]
May I get a second review for this change?
Or is it already OK to be pushed?
Thanks.

**Comment by dholmes-ora** [comment]
Two reviews needed.

**Comment by DamonFool** [comment]
Hi @pliden and @tschatzl ,

Could you please review this change?
Thanks.

**Review by pliden — APPROVED** [review]

**Comment by DamonFool** [comment]
Thanks @pliden .
/integrate

**Pull request closed without merging**
